File: app/src/main/java/com/sam_chordas/android/stockhawk/rest/YqlApi.java
package com.sam_chordas.android.stockhawk.rest;


import com.sam_chordas.android.stockhawk.model.HistoricalQuotesDataResponse;

import org.jetbrains.annotations.NotNull;

import kotlin.Metadata;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Query;

@Metadata(
        mv = {1, 1, 1},
        bv = {1, 0, 0},
        k = 1,
        d1 = {"\u0000\u001a\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0000\bf\u0018\u00002\u00020\u0001J\u0018\u0010\u0002\u001a\b\u0012\u0004\u0012\u00020\u00040\u00032\b\b\u0001\u0010\u0005\u001a\u00020\u0006H\'¨\u0006\u0007"},
        d2 = {"Lcom/sam_chordas/android/stockhawk/rest/YqlApi;", "", "getHistoricalQuotesData", "Lretrofit2/Call;", "Lcom/sam_chordas/android/stockhawk/model/HistoricalQuotesDataResponse;", "query", "", "production sources for module app"}
)
public interface YqlApi {
    @GET("yql?format=json&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys")
    @NotNull
    Call<HistoricalQuotesDataResponse> getHistoricalQuotesData(@Query("q") @NotNull String var1);
}

File: app/src/main/java/com/sam_chordas/android/stockhawk/rest/YqlApi.kt
package com.sam_chordas.android.stockhawk.rest

import android.support.annotation.IntRange
import com.github.aurae.retrofit2.LoganSquareConverterFactory
import com.sam_chordas.android.stockhawk.model.HistoricalQuotesDataResponse
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.logging.HttpLoggingInterceptor.Level.BODY
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.http.GET
import retrofit2.http.Query
import java.text.SimpleDateFormat
import java.util.*

private val YQL_ENDPOINT = "https://query.yahooapis.com/v1/public/"

val yqlApi: YqlApi = Retrofit.Builder()
        .baseUrl(YQL_ENDPOINT)
        .client(loggingClient())
        .addConverterFactory(LoganSquareConverterFactory.create())
        .build()
        .create(YqlApi::class.java)

private fun loggingClient() = OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor().setLevel(BODY))
            .build()

interface YqlApi {
    /**
     * @param query Call [buildHistoryQuery] to get the formatted query to pass here.
     */
    @GET("yql?format=json&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys")
    fun getHistoricalQuotesData(@Query("q") query: String): Call<HistoricalQuotesDataResponse>
}

fun YqlApi.getHistoryForMonths(symbol: String, @IntRange(from = 0) months: Int) =
        getHistoricalQuotesData(buildHistoryQuery(symbol, months))

fun buildHistoryQuery(symbol: String, @IntRange(from = 0) months: Int): String {
    val startDate = Calendar.getInstance().apply {
        add(Calendar.MONTH, -months)
    }
    return buildHistoryQuery(symbol, startDate.time, Date())
}

fun buildHistoryQuery(symbol: String, startDate: Date, endDate: Date): String {
    val startDateStr = dateFormatter.format(startDate)
    val endDateStr = dateFormatter.format(endDate)
    return buildHistoryQuery(symbol, startDateStr, endDateStr)
}

private fun buildHistoryQuery(symbol: String, startDate: String, endDate: String): String {
    val historyTable = "yahoo.finance.historicaldata"
    val symbolCondition = "symbol = \"$symbol\""
    val dateCondition = "startDate = \"$startDate\" and endDate = \"$endDate\""
    return "select * from $historyTable where $symbolCondition and $dateCondition"
}

private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)
